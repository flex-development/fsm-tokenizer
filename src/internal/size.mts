/**
 * @file Internal - size
 * @module vfile-tokenizer/internal/size
 */

import type { List } from '@flex-development/vfile-tokenizer'

/**
 * Get the size of a `list`.
 *
 * @internal
 *
 * @this {void}
 *
 * @param {List} list
 *  The list to measure
 * @return {number}
 *  Size of `list`
 */
function size(this: void, list: List): number {
  return Array.isArray(list) ? list.length : list.size
}

export default size
