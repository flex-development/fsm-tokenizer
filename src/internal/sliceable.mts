/**
 * @file Internal - sliceable
 * @module vfile-tokenizer/internal/sliceable
 */

/**
 * Check if `value` is an array or string.
 *
 * @internal
 *
 * @template {any} T
 *  Array item type
 *
 * @this {void}
 *
 * @param {unknown} value
 *  The thing to check
 * @return {value is T[] | string}
 *  `true` if `value` is an array or string, `false` otherwise
 */
function sliceable<T>(
  this: void,
  value: unknown
): value is T[] | string {
  return typeof value === 'string' || Array.isArray(value)
}

export default sliceable
