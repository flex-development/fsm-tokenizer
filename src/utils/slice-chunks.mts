/**
 * @file Utilities - sliceChunks
 * @module vfile-tokenizer/utils/sliceChunks
 */

import size from '#internal/size'
import sliceable from '#internal/sliceable'
import type { Chunk, List, Range } from '@flex-development/vfile-tokenizer'
import { ok } from 'devlop'

export default sliceChunks

/**
 * Get the chunks spanning `range`.
 *
 * @see {@linkcode Chunk}
 * @see {@linkcode List}
 * @see {@linkcode Range}
 *
 * @category
 *  utils
 *
 * @template {Chunk | string} T
 *  Chunk type
 *
 * @this {void}
 *
 * @param {List<T>} chunks
 *  List of chunks
 * @param {Range} range
 *  Position in stream
 * @return {T[]}
 *  List of chunks spanning `range`
 */
function sliceChunks<T extends Chunk | string>(
  this: void,
  chunks: List<T>,
  range: Range
): T[] {
  /**
   * Slice of chunks.
   *
   * @var {T[]} slice
   */
  let slice: T[] = []

  if (size(chunks)) {
    const { _bufferIndex: endBufferIndex, _index: endIndex } = range.end
    const { _bufferIndex: startBufferIndex, _index: startIndex } = range.start

    /**
     * List of chunks.
     *
     * @const {T[]} list
     */
    const list: T[] = [...chunks]

    if (startIndex === endIndex && startBufferIndex > 0 && endBufferIndex > 0) {
      ok(sliceable(list[startIndex]), 'expected buffer chunk')
      slice = [list[startIndex].slice(startBufferIndex, endBufferIndex) as T]
    } else {
      slice = list.slice(startIndex, endIndex)

      if (startBufferIndex > -1) {
        /**
         * First chunk in slice.
         *
         * @const {T | undefined} head
         */
        const head: T | undefined = slice.shift()

        if (sliceable(head)) slice.unshift(head.slice(startBufferIndex) as T)
      }

      if (endBufferIndex > 0) {
        ok(sliceable(list[endIndex]), 'expected buffer chunk')
        slice.push(list[endIndex].slice(0, endBufferIndex) as T)
      }
    }
  }

  return slice
}
