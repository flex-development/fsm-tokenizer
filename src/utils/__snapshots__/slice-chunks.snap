// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/sliceChunks > should return chunks spanning \`range\` (0) 1`] = `
[
  "--conditions",
]
`;

exports[`unit:utils/sliceChunks > should return chunks spanning \`range\` (1) 1`] = `
[
  "--conditions=",
  "vfile-tokenize",
]
`;

exports[`unit:utils/sliceChunks > should return chunks spanning \`range\` (2) 1`] = `
[
  [
    99,
    111,
    110,
    100,
    105,
    116,
    105,
    111,
    110,
    115,
  ],
]
`;

exports[`unit:utils/sliceChunks > should return chunks spanning \`range\` (3) 1`] = `
[
  [
    118,
    102,
    105,
    108,
    101,
    45,
    116,
    111,
    107,
    101,
    110,
    105,
    122,
    101,
    114,
  ],
]
`;
